### ISSUE TITLE
Tracking: Block AttnRes on Blackwell — from paper to measured result

### ISSUE THREAD

**Issue by sabdulmajid** [sabdulmajid/blackwell-switchyard#1]
**Tracking issue for the whole project.** Individual changes land as small draft PRs
linked from the table below. `PROJECT_STATE.md` in the repo carries the same status and the
decision log; this issue is the GitHub-side index.

## Goal

Answer, with reproducible measurements on RTX PRO 6000 Blackwell (`sm_120`):

> How expensive is Block AttnRes implemented conventionally, versus our best implementation?

Success is a **defensible performance story** — a correct operator, strong baselines, a
profiler-backed explanation of the bottleneck, and honest numbers. It is explicitly *not*
"a custom kernel exists".

## Ground rules this project holds itself to

- The baseline to beat is **`torch.compile`**, not eager. A win over naive eager alone is not a result.
- Every performance claim states GPU, software versions, dtype, exact shape, baseline, measurement
  method, repetition count, statistic, and whether compilation/allocation was included.
- Forward *and* backward both count. This is a training primitive.
- Temporary memory is a first-class metric, reported separately from memory the operator
  mathematically requires.
- Negative results are reported, not buried. If Inductor is already at the bandwidth roofline,
  that is the finding.
- No novelty claim survives without a literature and code search behind it.

## Milestones

| # | Milestone | State | PR |
|---|-----------|-------|----|
| 0 | Bootstrap + environment characterization | in progress | — |
| 1 | Faithful reference + correctness suite | not started | — |
| 2 | Benchmark & profiling infrastructure | not started | — |
| 3 | Strong framework baselines (eager / `torch.compile`) | not started | — |
| 4 | First optimized operator (evidence-selected) | not started | — |
| 5 | Backward / autograd optimization | not started | — |
| 6 | Blackwell-specific tuning | not started | — |
| 7 | Transformer integration + training smoke test | not started | — |
| 8 | Two-GPU validation | not started | — |
| 9 | Final benchmark study + technical report | not started | — |

## Machine

2 x NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition, CC 12.0, 188 SMs, 128 MiB L2,
95 GiB each, 300 W cap, PCIe-only between GPUs (no NVLink), driver 580.159.03.
Software: Ubuntu 24.04, torch 2.9.0+cu128, triton 3.5.0, CUDA toolkit 12.8 (nvcc 12.8.93),
Nsight Compute 2025.1.1, Nsight Systems 2024.6.2. Nothing was installed on the host.

## Attribution

Attention Residuals is the work of the Kimi Team at Moonshot AI
(arXiv:2603.15031, github.com/MoonshotAI/Attention-Residuals). This repository is an
independent performance-engineering implementation, not affiliated with Moonshot AI.

## Open questions for the owner

None yet. Technical choices are being decided by measurement.

**Comment by sabdulmajid** [comment]
> **Note on process:** the API token available to me has `contents:write` and
> `issues:write` but not `pull_requests:write`, so I can push branches but cannot
> open PRs. Branch descriptions are posted here as comments until that is
> granted; each one is ready to become a PR with a single click on the compare
> link. Granting **Pull requests: write** on the fine-grained PAT would let me
> open them directly.

---

# Branch: `machine-characterization` -> `main`

**Open this PR:** https://github.com/sabdulmajid/blackwell-switchyard/compare/main...machine-characterization?expand=1

Milestone 0. Establishes the numbers every later performance claim in this
repository divides by, and captures the constraints the host actually imposes.

Closes nothing yet; feeds #1.

## What is measured

Both RTX PRO 6000 Blackwell Max-Q (`sm_120`), agreeing to within 0.1%:

| Metric | GPU 0 | GPU 1 |
|---|---|---|
| DRAM copy | 1462 GB/s | 1462 GB/s |
| DRAM read | 1638 GB/s | 1639 GB/s |
| L2 peak copy | 6064 GB/s | 6066 GB/s |
| BF16 matmul (4096) | 299 TFLOP/s | 297 TFLOP/s |
| Triton launch overhead | 9.3 us | 9.5 us |
| CUDA-graph replay | 1.07 us | discarded (see below) |

Theoretical DRAM is 1792 GB/s (14001 MHz x 2 x 512-bit); measured read is 91% of
it. Nothing in this repo divides by the theoretical number.

## The bandwidth curve, and why it matters

| Working set | Copy GB/s |
|---|---|
| 4 MiB | 729 |
| 16 MiB | 2953 |
| 32 MiB | 5780 |
| 48 MiB | **6071** |
| 64 MiB | 5669 |
| 96 MiB | 1571 |
| >= 128 MiB | ~1462 |

L2 is **4.15x** DRAM bandwidth and holds 128 MiB. That is large enough to hold a
real tile of the target operator's inputs, which makes a two-pass formulation a
live candidate rather than an obvious loss. This shape drives the optimization
plan, so it is worth a reviewer's attention.

## Two measurement bugs found and fixed

Flagging these because they are the kind of thing that silently produces a
flattering result:

1. **Launch overhead masquerading as bandwidth.** The first version timed one
   kernel per working-set size. At 32 MiB the ~10 us launch cost dominated, and
   the probe reported L2 as *slower* than DRAM (1306 vs 1458 GB/s) -- the cache
   hierarchy inverted. Fixed by repeating each kernel enough times to move a
   fixed total volume at every size, so overhead amortizes identically.
2. **An empty CUDA graph replaying in nanoseconds.** Capture silently fails on
   GPU 1 and PyTorch only emits a warning. Reported as 0.06 us/op, a 100x
   "speedup" over GPU 0. The probe now verifies capture recorded the expected
   op count and discards the measurement otherwise, which is why GPU 1 has no
   number in the table above rather than a great one.

## Constraints this imposes on the project

- **Nsight Compute is unusable.** `ncu` runs but counter collection fails with
  `ERR_NVGPUCTRPERM`. Lifting it means setting
  `NVreg_RestrictProfilingToAdminUsers` host-wide, which this project will not
  do on a shared machine. Substitutes: `nsys` (verified working),
  `torch.profiler`, and analytical roofline -- viable here because the target
  operator is memory-bound by ~two orders of magnitude, so achieved bandwidth
  from latency plus an exact byte count answers most of what ncu would.
- **No NVLink.** Peer access is enabled but runs over PCIe Gen4 at 25.8 GB/s
  measured. Any two-GPU work must assume a slow link.
- **No `python3.12-dev` on the host**, so no `Python.h`. Triton and Inductor
  both build a C launcher shim per kernel, so `torch.compile` does not work at
  all without it. `scripts/fetch_python_headers.sh` unpacks the matching Ubuntu
  package into a gitignored `.local-toolchain/` and `scripts/env.sh` points the
  compiler at it. Nothing is installed; no system path is written to; no sudo.

## Host safety

No drivers, toolkits, kernel modules, system packages, power limits, or clocks
were touched. No processes other than this project's were run on the GPUs. Both
cards were idle before and after.

## How to reproduce

```bash
source scripts/env.sh          # after scripts/fetch_python_headers.sh
python bench/machine.py        # writes results/machine.json
python scripts/render_machine.py   # regenerates docs/machine.md
```

## Open for review

- Bus width (512-bit) is an assumption recorded as such in the JSON, not a
  measurement -- `nvidia-smi` does not expose it. It affects only the
  theoretical-bandwidth line, which nothing depends on.
- CUDA-graph capture failing on GPU 1 is unexplained. It does not block
  anything (identical hardware, GPU 0 measured), so I recorded it and moved on
  rather than chasing it.

**Comment by sabdulmajid** [comment]
# Branch: `fused-operator` -> `main`

**Open this PR:** https://github.com/sabdulmajid/blackwell-switchyard/compare/main...fused-operator?expand=1

Depends on `machine-characterization` only for `scripts/env.sh` (needed to run anything —
the host has no `python3.12-dev`, so Triton and Inductor cannot build launcher shims).
Reviewable independently; the code does not otherwise overlap.

Milestones 1-5 and most of 10.

## Result

RTX PRO 6000 Blackwell (`sm_120`), bf16, `N=9 B=1 T=4096 D=2048`. Median of 100
CUDA-event-timed reps, 128 MiB L2 flushed between them, steady state after compilation.

| | best framework baseline | switchyard | |
|---|---|---|---|
| forward | 0.430 ms | **0.123 ms** | **3.50x** |
| forward + backward | 1.918 ms | **0.362 ms** | **5.29x** |
| kernels (fwd / fwd+bwd) | 6 / 16 | **1 / 4** | |
| workspace above the data | 592 MiB | **464 MiB** | -22% |

The baseline is the fastest of six framework configurations (two algebraic formulations x
eager plus three `torch.compile` modes), not eager PyTorch.

Forward sustains **94-101% of speed of light** for N from 2 to 32 — a kernel touching the
same bytes with the same access pattern and launch geometry, minus the softmax. The forward
is done; I would rather say that than imply more is available.

## What is here

- `reference.py` — paper-faithful, float64 oracle, two-phase online-softmax helpers
- `baselines.py` — `paper_form` and `folded_form`
- `triton_op.py` — two forward and two backward strategies, dispatched on measured tile budget
- `bench/harness.py`, `bench/bench_operator.py` — oracle check before timing, L2 flush,
  order statistics, kernel counts, workspace memory, speed-of-light ceiling
- 99 tests; `docs/report.md`; `docs/results.md` regenerated from raw JSON

## Things I would want a reviewer to push on

**The dispatch threshold was reasoned about first, and the reasoning was wrong.** A
register-pressure estimate put the resident tile limit at 16384 fp32 values; sweeping it
gave 32768. Sitting exactly at the difference is `N=9, D=2048` — the most representative
shape, since the paper uses ~8 blocks plus the embedding — which was running at 616 GB/s
instead of 1365 GB/s. A test now pins it. Two other hand-picked constants were also wrong:
`num_stages=1` beats 2-3 nearly everywhere, and the tiled kernel wants `BLOCK_D=2048`, not
the small tile a reduction suggests.

**The backward fallback was a regression, not just an absence.** Shapes outside the resident
budget fell back to autograd, but the fallback recomputed the forward before differentiating,
so fwd+bwd ran at **0.51x** the baseline at `N=9, D=4096` — an ordinary model shape. The
tiled backward (two kernels, because `dw` reduces over tokens while the softmax statistics
need a full-`D` reduction per token) fixes it to 4.40x. Worth checking I have not left a
similar hole elsewhere.

**`torch.compile` silently falls back at N >= 16.** `paper_compiled` returns latencies and
kernel counts bit-identical to `paper_eager` — Inductor declines to fuse, no error. Flagging
because it inflates our margin at exactly those shapes; the tables use whichever framework
config was actually fastest, so `folded_eager` becomes the baseline there.

**The kernel is more accurate than the baseline, which is worth double-checking.** 1.00x the
bf16 rounding floor against 1.6-3.3x for eager, because it accumulates in fp32 and rounds
once. If that looks too good, the oracle comparison is in `bench/harness.py::check_against`
and the floor is computed per-shape.

**Measurement bugs I found and fixed** (both would have flattered us, both documented in
`docs/report.md`): a bandwidth probe reporting L2 as *slower* than DRAM because launch
overhead dominated at small working sets; and benchmark inputs drawn as `w ~ N(0, 0.5^2)`,
which puts the logit standard deviation at ~16 and saturates the softmax into a regime no
trained model occupies — every implementation "failed" correctness until the inputs were
made physical.

## Not done

- Head-to-head against the existing third-party kernels (Liger, fla, catswe) on this
  hardware. In progress. The README's framing depends on it.
- End-to-end Transformer integration and two-GPU validation. In progress / next.
- fp16 and fp32 are tested for correctness but not swept for performance.

## Validation

`python -m pytest tests/ -q` -> 99 passed. Every benchmarked implementation is checked
against a float64 oracle before it is timed; failures are excluded, not annotated. No host
configuration was changed.

**Comment by sabdulmajid** [comment]
## Two token permissions needed

Neither blocks progress — work is continuing — but both need an action only you can take.

**1. `Pull requests: write`** — I can push branches and file issues but cannot open PRs.
Branch descriptions are posted here as comments with one-click compare links.

**2. `workflow`** — I cannot push anything under `.github/workflows/`. Push is rejected with:

```
refusing to allow a Personal Access Token to create or update workflow
`.github/workflows/ci.yml` without `workflow` scope
```

The CI workflow below is written and verified locally (`pytest` on CPU: 22 passed,
1 skipped; `ruff check src bench scripts tests`: all checks passed). It is deliberately
modest — a GitHub runner has no Blackwell GPU, so it validates only what is genuinely
machine-independent and makes no attempt to measure performance. Performance is gated
locally by `scripts/check_regression.py` instead.

Add it via the GitHub web UI (Actions → new workflow) or grant the scope and I will push it.

<details>
<summary><code>.github/workflows/ci.yml</code></summary>

```yaml
name: ci

# What this can and cannot check.
#
# GitHub runners have no GPU, let alone a Blackwell one, so this workflow makes
# no attempt to validate performance. Doing so would be theatre. It runs the
# parts that are genuinely machine-independent: the reference implementation's
# mathematics, its gradients, and lint.
#
# Performance is gated locally instead, by scripts/check_regression.py against a
# stored baseline. See docs/report.md.

on:
  push:
    branches: [main]
  pull_request:
  workflow_dispatch:

jobs:
  reference:
    name: reference semantics (CPU)
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - uses: actions/setup-python@v5
        with:
          python-version: "3.12"

      - name: Install CPU-only PyTorch
        run: |
          python -m pip install --upgrade pip
          pip install --index-url https://download.pytorch.org/whl/cpu torch
          pip install pytest numpy

      - name: Reference tests
        # test_triton_op.py skips itself without CUDA. What remains is the
        # oracle: paper semantics, float64 gradcheck and gradgradcheck, and the
        # online-softmax merge used by the paper's two-phase schedule. If these
        # pass, every GPU comparison is being made against a correct reference.
        run: python -m pytest tests/ -q
        env:
          PYTHONPATH: src

  lint:
    name: lint
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v5
        with:
          python-version: "3.12"
      - run: pip install ruff
      - run: ruff check src bench scripts tests

```
</details>

**Comment by sabdulmajid** [comment]
# Branch update: `fused-operator` — third-party comparison, model integration, two GPUs

**Open this PR:** https://github.com/sabdulmajid/blackwell-switchyard/compare/main...fused-operator?expand=1

Milestones 7-10 landed since the previous comment. 113 tests pass, lint clean.

## The end-to-end number

1.3B decoder, batch 4 x seq 2048, bf16 ([`docs/model.md`](../blob/fused-operator/docs/model.md)):

| variant | step ms | tokens/s | peak GiB | mechanism costs |
|---|---|---|---|---|
| standard PreNorm (control) | 428.89 | 19101 | 30.33 | -- |
| Block AttnRes, framework | 704.54 | 11627 | 31.87 | **39% of the step** |
| **Block AttnRes, switchyard** | **484.21** | **16918** | **31.86** | **11% of the step** |

Fusing takes the mechanism from 39% of a step to 11% — 3.4x less overhead, 1.46x end to
end. Adopting Block AttnRes then costs 11% of step time and 1.5 GiB (5%) over a standard
residual. The end-to-end figure is smaller than the operator's 3.5-5x because the operator
is one part of a step; I quote the Amdahl arithmetic rather than the kernel speedup.

Separately: the paper's pseudocode `torch.stack`s at every site, which costs 265 slab
copies per forward and 6.75 GiB. Measured and reported rather than quietly avoided.

## The head-to-head, which does not go entirely our way

[`docs/third_party.md`](../blob/fused-operator/docs/third_party.md). All four verified
against the same float64 oracle before timing, with each project's RMSNorm gain, logit
scale and output-norm set so they compute the same function.

- **Forward is a four-way tie at the memory ceiling.** switchyard, fla and catswe are all
  within a few percent of speed of light. fla is *ahead* of us at D=8192.
- **fwd+bwd: we take 7 of 10 shapes, Liger takes 3** — every one at D>=4096 or N=32, which
  is exactly where our tiled backward runs instead of the resident one. Real deficiency.
- **catswe is 4.3x faster on batched pseudo-queries** (0.215 ms vs our 0.930 ms for 8).
  Their two-phase schedule reads the sources once per block; ours once per call. An 8x
  traffic difference no kernel tuning closes. This is the largest thing left, and it is an
  API problem, not a kernel one.

## Two GPUs

| variant | 1 GPU | 2 GPU | speedup | efficiency |
|---|---|---|---|---|
| standard | 17555 | 29799 | 1.70x | 85% |
| switchyard | 16238 | 28138 | **1.73x** | **87%** |

Gradients bit-identical across ranks after all-reduce (max deviation 0.0e+00) — the check
that matters, since a custom autograd returning a wrong gradient for one input passes every
single-GPU test and still corrupts a distributed run.

## The thing I most want reviewed

**I nearly published a number that flattered us by 4.3x.** fla takes a sequence of source
tensors, which is how a real model holds them. My first harness handed it `N` views of one
leaf, so autograd routed every gradient through `N` slice-backwards. It reported fla as 11x
slower than us at N=9 and 29x at N=32; the true figures are 2.7x and 1.5x. Nothing in the
harness would have caught it — only asking "why would a maintained kernel be *that* bad?"
did. Both numbers are kept in the JSON because the gap is itself the cost of a pointer-table
API when sources do live in one buffer. Written up in `docs/third_party.md`.

Also worth a look: the arena in `model.py` writes sources into one buffer in place, hands
the kernel an alias with its own version counter, and restores the partial-sum slot during
backward because later sublayers overwrite it. It is delicate. I validated it numerically
rather than by reading — forward bit-identical to plain stacking, gradients agreeing to
1e-8 (accumulation order) — but a second pair of eyes on `_ArenaAttnRes` would be welcome.

## What is left, in order of measured value

1. A batched-query API (the catswe gap above).
2. Improve the tiled backward (the Liger gap above).
3. fp16/fp32 performance sweeps; currently correctness-tested only.

**Comment by sabdulmajid** [comment]
Status on 2026-09-06:

- The development branch and the immutable campaign branch are at 28f6b83.
- The complete CPU test suite passes: 245 passed and 3 skipped.
- The offline sm_120 contract now covers 90 Triton specializations and 16 CUDA kernel instances.
- Every experimental candidate has zero forbidden compiler spills. The contract also records 21 spill-prone specializations in the accepted baseline. This is direct evidence for the backward-path bottleneck.
- The guarded GPU campaign is armed, but it has not started a benchmark attempt. It requires both GPUs to stay idle for 30 minutes. It checks once per minute while it waits. It stops our workload if another process appears on the selected GPU. It allows at most three attempts within 36 hours.

No new GPU performance claim is accepted yet. The next update will use the committed raw campaign evidence.
